src/sinks/s3_common/sink.rs

```diff
@@ -1,68 +1,221 @@
-use std::{fmt, num::NonZeroUsize};
+use std::{fmt, num::NonZeroUsize, sync::Arc};
 
 use async_trait::async_trait;
+use bytes::{Bytes, BytesMut};
+use codecs::encoding::{Framer, Serializer};
 use futures::stream::BoxStream;
 use futures_util::StreamExt;
+use tokio_util::codec::Encoder as _;
 use tower::Service;
-use vector_common::request_metadata::MetaDescriptive;
+use vector_common::request_metadata::{GroupedCountByteSize, MetaDescriptive, RequestMetadata};
 use vector_core::{
+    config::telemetry,
     event::Finalizable,
+    partition::Partitioner,
     sink::StreamSink,
     stream::{BatcherSettings, DriverResponse},
+    ByteSizeOf, EstimatedJsonEncodedSizeOf,
 };
 
-use crate::internal_events::SinkRequestBuildError;
 use crate::{
+    codecs::Transformer,
     event::Event,
-    sinks::util::{RequestBuilder, SinkBuilderExt},
+    sinks::util::{encoding::write_all, Compression, SinkBuilderExt},
 };
+use crate::{internal_events::SinkRequestBuildError, sinks::util::Compressor};
 
-use super::partitioner::{S3KeyPartitioner, S3PartitionKey};
-
-pub struct S3Sink<Svc, RB> {
+pub struct S3Sink<Svc, RB, Part> {
     service: Svc,
     request_builder: RB,
-    partitioner: S3KeyPartitioner,
+    partitioner: Part,
+    transformer: Transformer,
+    serializer: Serializer,
+    framer: Framer,
     batcher_settings: BatcherSettings,
 }
 
-impl<Svc, RB> S3Sink<Svc, RB> {
+impl<Svc, RB, Part> S3Sink<Svc, RB, Part> {
     pub const fn new(
         service: Svc,
         request_builder: RB,
-        partitioner: S3KeyPartitioner,
+        partitioner: Part,
+        transformer: Transformer,
+        serializer: Serializer,
+        framer: Framer,
         batcher_settings: BatcherSettings,
     ) -> Self {
         Self {
             partitioner,
             service,
             request_builder,
+            transformer,
+            serializer,
+            framer,
             batcher_settings,
         }
     }
 }
 
-impl<Svc, RB> S3Sink<Svc, RB>
+pub struct EncodedEvent {
+    pub original: Event,
+    pub json_size: GroupedCountByteSize,
+    pub encoded: BytesMut,
+}
+
+impl Finalizable for EncodedEvent {
+    fn take_finalizers(&mut self) -> vector_common::finalization::EventFinalizers {
+        // TODO: Do we need to worry about finalizer getting cloned prior to encoding?
```

Comment:
Is this still a TODO? Where does it get the extra clone, are you referring to when the parent event is cloned?